### PR TITLE
Support for displaying display items based on conditions.

### LINKED ISSUE #1835: SDC | enableWhenExpression not working with displayCategory Extension
**Describe the bug**
In a questionnaire item with enableWhenExpression & displayCategory Extension it is not executing correctly.

**Component**
SDC Library

**To Reproduce**
1.Replace the new-paginated-registration-paginated.json with this questionnaire :
[displayCategoryEnableWhenQuestionnaire.txt](https://github.com/google/android-fhir/files/10551783/displayCategoryEnableWhenQuestionnaire.txt)

**Expected behavior**
in the questionnaire "with wheezing" instruction should be displayed.

**Screenshots**
Current output:
Screenshot from 2023-01-31 14-58-19

Expected output:
Screenshot from 2023-01-31 15-06-20

cc: @jingtang10 @delcroip

### PR DESCRIPTION
**Pull request by MJ1998**
**IMPORTANT: All PRs must be linked to an issue (except for extremely trivial and straightforward changes).**

Fixes #1835 

**Description**
Support for items of display type given some conditions (defined by enableWhen, enableWhenExpression) by passing "enabledDisplayItems" information in QuestionnaireViewItem.

**Type**
Choose one: Bug fix

**Checklist**
- [x] I have read and acknowledged the [Code of conduct](https://github.com/google/android-fhir/blob/master/docs/CODE_OF_CONDUCT.md).
- [x] I have read the [Contributing](https://github.com/google/android-fhir/wiki/Contributing) page.
- [x] I have signed the Google [Individual CLA](https://cla.developers.google.com/about/google-individual), or I am covered by my company's [Corporate CLA](https://cla.developers.google.com/about/google-corporate ).
- [x] I have discussed my proposed solution with code owners in the linked issue(s) and we have agreed upon the general approach.
- [x] I have run `./gradlew spotlessApply` and `./gradlew spotlessCheck` to check my code follows the style guide of this project.
- [x] I have run `./gradlew check` and `./gradlew connectedCheck` to test my changes locally.
- [x] I have built and run the demo app(s) to verify my change fixes the issue and/or does not break the demo app(s).

### CODE REVIEW

File: datacapture/src/main/java/com/google/android/fhir/datacapture/views/HeaderView.kt [hunk 1]
```diff
@@ -47,11 +46,13 @@ internal class HeaderView(context: Context, attrs: AttributeSet?) : LinearLayout
...
-  fun bind(questionnaireItem: Questionnaire.QuestionnaireItemComponent) {
-    prefix.updateTextAndVisibility(questionnaireItem.localizedPrefixSpanned)
-    question.updateTextAndVisibility(questionnaireItem.localizedTextSpanned)
-    hint.updateTextAndVisibility(questionnaireItem.localizedInstructionsSpanned)
-    initHelpButton(this, questionnaireItem)
+  fun bind(questionnaireViewItem: QuestionnaireViewItem) {
+    prefix.updateTextAndVisibility(questionnaireViewItem.questionnaireItem.localizedPrefixSpanned)
+    question.updateTextAndVisibility(questionnaireViewItem.questionnaireItem.localizedTextSpanned)
+    hint.updateTextAndVisibility(
+      questionnaireViewItem.enabledDisplayItems?.localizedInstructionsSpanned
```
Comment (jingtang10) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/views/HeaderView.kt:53]: use `let`

```
questionnaireViewItem.questionnaireItem.let {
  prefix.update...(it.prefix)
  ...
}
```

File: datacapture/src/main/java/com/google/android/fhir/datacapture/views/QuestionnaireViewItem.kt [hunk 2]
```diff
@@ -78,7 +78,8 @@ data class QuestionnaireViewItem(
     {
       emptyList()
     },
-  internal val draftAnswer: Any? = null
+  internal val draftAnswer: Any? = null,
+  internal val enabledDisplayItems: List<Questionnaire.QuestionnaireItemComponent>? = null
```
[the remark this answers is not in the document]
Comment (jingtang10) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/views/QuestionnaireViewItem.kt:82]: can this default to empty list?

**Review by jingtang10 — CHANGES_REQUESTED** [review]

File: datacapture/src/main/java/com/google/android/fhir/datacapture/views/QuestionnaireViewItem.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (MJ1998) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/views/QuestionnaireViewItem.kt]: Done

File: datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireViewModel.kt [hunk 3]
```diff
@@ -627,7 +627,13 @@ internal class QuestionnaireViewModel(application: Application, state: SavedStat
             answersChangedCallback = answersChangedCallback,
             resolveAnswerValueSet = { resolveAnswerValueSet(it) },
             resolveAnswerExpression = { resolveAnswerExpression(it) },
-            draftAnswer = draftAnswerMap[questionnaireResponseItem]
+            draftAnswer = draftAnswerMap[questionnaireResponseItem],
+            enabledDisplayItems =
+              questionnaireItem.item.filter {
+                it.type == Questionnaire.QuestionnaireItemType.DISPLAY &&
+                  (it.isInstructionsCode || it.isFlyoverCode || it.isHelpCode) &&
```
[the remark this answers is not in the document]
Comment (MJ1998) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/QuestionnaireViewModel.kt]: Added QuestionnaireItemComponent.isDisplayItem in MoreQuestionnaireItemComponents.kt

File: datacapture/src/main/java/com/google/android/fhir/datacapture/views/HeaderView.kt (hunk 1 above)
Comment (MJ1998) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/views/HeaderView.kt]: Done

File: datacapture/src/main/java/com/google/android/fhir/datacapture/extensions/MoreQuestionnaireItemComponents.kt [hunk 4]
```diff
@@ -322,9 +312,11 @@ internal val Questionnaire.QuestionnaireItemComponent.localizedInstructionsSpann
  * present) is used as the fly-over text of the parent question.
  */
 internal val Questionnaire.QuestionnaireItemComponent.localizedFlyoverSpanned: Spanned?
+  get() = item.localizedFlyoverSpanned
+
+internal val List<Questionnaire.QuestionnaireItemComponent>.localizedFlyoverSpanned: Spanned?
```
[the remark this answers is not in the document]
Comment (MJ1998) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/extensions/MoreQuestionnaireItemComponents.kt:317]: Done

File: datacapture/src/main/java/com/google/android/fhir/datacapture/views/HeaderView.kt [hunk 5]
```diff
@@ -47,12 +46,16 @@ internal class HeaderView(context: Context, attrs: AttributeSet?) : LinearLayout
   private var hint: TextView = findViewById(R.id.hint)
   private var errorTextView: TextView = findViewById(R.id.error_text_at_header)
 
-  fun bind(questionnaireItem: Questionnaire.QuestionnaireItemComponent) {
-    prefix.updateTextAndVisibility(questionnaireItem.localizedPrefixSpanned)
-    question.updateTextAndVisibility(questionnaireItem.localizedTextSpanned)
-    hint.updateTextAndVisibility(questionnaireItem.localizedInstructionsSpanned)
-    initHelpButton(this, questionnaireItem)
-    //   Make the entire view GONE if there is nothing to show. This is to avoid an empty row in the
+  fun bind(questionnaireViewItem: QuestionnaireViewItem) {
```
[the remark this answers is not in the document]
Comment (MJ1998) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/views/HeaderView.kt:49]: Jing suggested to use `let` but personally even I think GroupHeaderView is neater.
Both bind functions and HeaderViews look similar now!

File: datacapture/src/main/java/com/google/android/fhir/datacapture/extensions/MoreQuestionnaireItemComponents.kt [hunk 6]
```diff
@@ -335,9 +327,11 @@ internal val Questionnaire.QuestionnaireItemComponent.localizedFlyoverSpanned: S
  * code is used as the instructions of the parent question.
  */
 internal val Questionnaire.QuestionnaireItemComponent.localizedHelpSpanned: Spanned?
+  get() = item.localizedHelpSpanned
+
+internal val List<Questionnaire.QuestionnaireItemComponent>.localizedHelpSpanned: Spanned?
```
[the remark this answers is not in the document]
Comment (MJ1998) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/extensions/MoreQuestionnaireItemComponents.kt:332]: Done

File: datacapture/src/main/java/com/google/android/fhir/datacapture/extensions/MoreQuestionnaireItemComponents.kt [hunk 7]
```diff
@@ -308,9 +296,11 @@ val Questionnaire.QuestionnaireItemComponent.localizedPrefixSpanned: Spanned?
  * code is used as the instructions of the parent question.
  */
 internal val Questionnaire.QuestionnaireItemComponent.localizedInstructionsSpanned: Spanned?
+  get() = item.localizedInstructionsSpanned
+
+internal val List<Questionnaire.QuestionnaireItemComponent>.localizedInstructionsSpanned: Spanned?
```
[the remark this answers is not in the document]
Comment (MJ1998) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/extensions/MoreQuestionnaireItemComponents.kt:303]: Done.

File: datacapture/src/main/java/com/google/android/fhir/datacapture/views/GroupHeaderView.kt [hunk 8]
```diff
@@ -37,19 +36,22 @@ internal class GroupHeaderView(context: Context, attrs: AttributeSet?) :
...
+  private val question = findViewById<TextView>(R.id.question)
+  private val hint = findViewById<TextView>(R.id.hint)
+
+  fun bind(questionnaireViewItem: QuestionnaireViewItem) {
     initHelpViews(
       helpButton = findViewById(R.id.helpButton),
       helpCardView = findViewById(R.id.helpCardView),
       helpTextView = findViewById(R.id.helpText),
-      questionnaireItem
+      questionnaireViewItem.questionnaireItem
```
Comment (omarismail94) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/views/GroupHeaderView.kt:48]: if you are listing the parameter names of the variables of the function, list them all:
```suggestion
      questionnaireItem = questionnaireViewItem.questionnaireItem
```

**Review by omarismail94 — COMMENTED** [review]
Could you add an espresso test that loads a quesitonnaire similar to the [one]( https://github.com/google/android-fhir/files/10551783/displayCategoryEnableWhenQuestionnaire.txt) posted in the issue, asserting the right text is shown?

File: datacapture/src/main/java/com/google/android/fhir/datacapture/views/GroupHeaderView.kt (hunk 8 above)
Comment (MJ1998) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/views/GroupHeaderView.kt]: Done. Added `displayItems_shouldGetEnabled_withAnswerChoice` in `QuestionnaireUiEspressoTest`.

File: datacapture/sampledata/questionnaire_with_enabled_display_items.json [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+            }
+          ],
+          "linkId": "1.2",
+          "text": "Text when yes is selected",
+          "type": "display"
+        }
+      ]
+    }
+  ]
+}
```
Comment (omarismail94) [inline datacapture/sampledata/questionnaire_with_enabled_display_items.json:66]: nit: add blank line

File: datacapture/src/androidTest/java/com/google/android/fhir/datacapture/QuestionnaireUiEspressoTest.kt [hunk 10]
```diff
@@ -361,6 +363,41 @@ class QuestionnaireUiEspressoTest {
...
+    onView(withId(R.id.hint)).check { view, _ ->
+      val hintVisibility = (view as TextView).visibility
+      assertThat(hintVisibility).isEqualTo(View.GONE)
+    }
+
+    onView(withId(R.id.yes_radio_button)).perform(ViewActions.click())
+
+    onView(withId(R.id.hint)).check { view, _ ->
+      val hintVisibility = (view as TextView).visibility
+      val hintText = (view as TextView).text.toString()
```
Comment (omarismail94) [inline datacapture/src/androidTest/java/com/google/android/fhir/datacapture/QuestionnaireUiEspressoTest.kt:379]:
```suggestion
      val hintText = view.text.toString()
```

File: datacapture/src/androidTest/java/com/google/android/fhir/datacapture/QuestionnaireUiEspressoTest.kt [hunk 11]
```diff
@@ -361,6 +363,41 @@ class QuestionnaireUiEspressoTest {
...
+      assertThat(hintVisibility).isEqualTo(View.VISIBLE)
+      assertThat(hintText).isEqualTo("Text when yes is selected")
+    }
+
+    onView(withId(R.id.no_radio_button)).perform(ViewActions.click())
+
+    onView(withId(R.id.hint)).check { view, _ ->
+      val hintVisibility = (view as TextView).visibility
+      val hintText = (view as TextView).text.toString()
```
Comment (omarismail94) [inline datacapture/src/androidTest/java/com/google/android/fhir/datacapture/QuestionnaireUiEspressoTest.kt:388]:
```suggestion
      val hintText = view.text.toString()
```

**Review by omarismail94 — APPROVED** [review]
Great job! Just resolve these minor comments and then merge

**Comment by omarismail94** [comment]
@jingtang10 resolving your reviews as @MJ1998 completed them!

**Pull request merged**